spec.emu

```diff
@@ -835,12 +1757,262 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
+      </ul>
+
+      <emu-clause id="sec-get-%abstractmodulesourceprototype%.@@tostringtag">
+        <h1>get %AbstractModuleSourcePrototype% [ @@toStringTag ]</h1>
+        <p>%AbstractModuleSourcePrototype%`[@@toStringTag]` is an accessor property whose set accessor function is *undefined*. Its get accessor function performs the following steps when called:</p>
+        <emu-alg>
+          1. Let _O_ be the *this* value.
+          1. If _O_ is not an Object, return *undefined*.
+          1. If _O_ does not have a [[ModuleSourceRecord]] internal slot, return *undefined*.
+          1. Let _name_ be _O_.[[ModuleSourceRecord]].[[SourseClassName]]
+          1. Assert: _name_ is a String.
+          1. Return _name_.
+        </emu-alg>
+        <p>This property has the attributes { [[Enumerable]]: *false*, [[Configurable]]: *true* }.</p>
+        <p>The initial value of the *"name"* property of this function is *"get [Symbol.toStringTag]"*.</p>
+      </emu-clause>
+    </emu-clause>
+
+    <emu-clause id="sec-modulesource-constructor">
+      <h1>The ModuleSource Constructor</h1>
+      <p>The ModuleSource Object represents the source phase of a SourceTextModuleRecord.</p>
+
+      <p>The ModuleSource constructor:</p>
+      <ul>
+        <li>is <dfn>%ModuleSource%</dfn>.</li>
+        <li>is the initial value of the *"ModuleSource"* property of the global object.</li>
+        <li>has a [[Prototype]] internal slot whose value is %AbstractModuleSourcePrototype%.</li>
+        <li>does not have a [[ModuleSourceRecord]] internal slot.
+      </ul>
+    </emu-clause>
+
+    <emu-clause id="sec-properties-of-the-modulesource-prototype-object">
+      <h1>Properties of the ModuleSource Prototype Object</h1>
+      <p>The <dfn>ModuleSource prototype object</dfn>:</p>
+      <ul>
+        <li>is <dfn>%ModuleSource.prototype%</dfn>.</li>
+        <li>has a [[Prototype]] internal slot whose value is %AbstractModuleSourcePrototype%.</li>
+        <li>is not a ModuleSource instance or a %AbstractModuleSourcePrototype% and does not have a [[ModuleSourceRecord]] internal slot.
+      </ul>
+    </emu-clause>
+
+    <emu-clause id="sec-properties-of-modulesource-instances">
+      <h1>Properties of ModuleSource Instances</h1>
+      <p>ModuleSource instances are ordinary objects that inherit properties from the ModuleSource Prototype.</p>
+      <p>ModuleSource instances have a [[ModuleSourceRecord]] internal slot containing their associated ModuleSource Record whose [[SourceClassName]] field is *"ModuleSource"*.</p>
+
+      <emu-note>
+        <p>Only SourceTextModuleRecord modules may be represented in their source phase by a ModuleSource instance and the *"ModuleSource"* string value for [[SourceClassName]].</p>
+      </emu-note>
+    </emu-clause>
+
+    <emu-clause id="sec-createmodulesource" type="abstract operation">
+      <h1>
+        CreateModuleSource (
+          _sourceClassName_: a String,
+          _hostDefined_: anything,
```

Comment:
```suggestion
          _hostDefined_: unknown,
```

Reply:
ECMA-262 does use the term `anything` for `_hostDefined_` on scripts and modules in the spec currently. This would need to be a whole spec edit to change I think.

Reply:
ah ok, if it's already used then obv we'd stick with it here :-)